### PR TITLE
Add/remove trainer and testing

### PR DESCRIPTION
**Pull request by yann300**
- allow adding/removing trainers
 - trainers can provide a list of addresses that will be allowed to mint a remixer.

### CODE REVIEW

File: tests/test.ts [hunk 1]
```diff
@@ -10,9 +10,17 @@ import { expect } from "chai";
 let remix: ethers.Contract
 let proxy: ethers.Contract
 let verifier: ethers.Contract
+
+let owner, betatester, user, betatester2, user2, user3, trainer1, student1
+
+describe("Basic remix reward deploy", function () {
```
Comment (Aniket-Engg) [inline tests/test.ts:16]: I think this describe text should be updated

File: tests/test.ts [hunk 2]
```diff
@@ -156,25 +152,68 @@ describe("Basic remix reward deploy", function () {
...
   });
-});
+})
+
+describe("Add / remove trainer", function () {
+  it("Add a trainer should fail", async function () {
+    await expect(remix.connect(trainer1).addTrainer(trainer1.address)).to.be.revertedWith('account 0x5c6b0f7bf3e7ce046039bd8fabdfd3f9f5021678 is missing role 0x0000000000000000000000000000000000000000000000000000000000000000')        
+  })
+
+  it("Add a trainer should success", async function () {
```
Comment (Aniket-Engg) [inline tests/test.ts:176]: There should be one more test case to check that trainer can be added by owner only not by any other address

File: contracts/RemixRewardUpgradable.sol [hunk 3]
```diff
@@ -37,12 +37,19 @@ contract Remix is Initializable, ERC721Upgradeable, ERC721EnumerableUpgradeable,
...
+    mapping(address => uint) public trainers;
+
     struct TokenData {
         string payload;
         string tokenType;
         bytes hash;
     }
 
+    modifier isTrainer() {
+        require(trainers[msg.sender] == 1, "Caller is not a trainer");
```
Comment (Aniket-Engg) [inline contracts/RemixRewardUpgradable.sol:49]: I think `===` operator can be used

File: contracts/RemixRewardUpgradable.sol [hunk 4]
```diff
@@ -129,7 +136,15 @@ contract Remix is Initializable, ERC721Upgradeable, ERC721EnumerableUpgradeable,
...
-    function grantRemixersMinting (address[] calldata remixers, uint amount) public onlyRole(DEFAULT_ADMIN_ROLE)  {
+    function addTrainer (address trainer) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        trainers[trainer] = 1;
+    }
+
+    function removeTrainer (address trainer) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        delete trainers[trainer];
+    }
+
+    function grantRemixersMinting (address[] calldata remixers, uint amount) public isTrainer()  {
```
Comment (Aniket-Engg) [inline contracts/RemixRewardUpgradable.sol:147]: With this update, admin will not be able to call this method. Is that intentional?

Comment (yann300) [reply contracts/RemixRewardUpgradable.sol:147]: yes I did that to make the function as light as possible. An admin can just add himself as trainer.

File: tests/test.ts (hunk 2 above)
Comment (yann300) [reply tests/test.ts:176]: I think this is done in "Add a trainer should fail"

File: contracts/RemixRewardUpgradable.sol (hunk 3 above)
Comment (yann300) [reply contracts/RemixRewardUpgradable.sol:49]: not in solidity.

File: README.md [hunk 5]
```diff
@@ -1,46 +1,35 @@
...
-```
+Remix Project rewards contributors, beta testers, and UX research participants with NFTs deployed on Optimism.
+Remix Reward holders are able to mint a second “Remixer” user NFT badge to any wallet address of their choice.
+This feature is a way to reward Remix contributors to help grow our user base into a larger and more genuine open source community of practice.
 
-# Etherscan verification
+Remix Rewards are currently not transferable. This feature leaves open the future possibility of granting holders proportional voting power to help the community decide on new features for the IDE and/or other issues governing the development of the Remix toolset.
 
-To try out Etherscan verification, you first need to deploy a contract to an Ethereum network that's supported by Etherscan, such as Ropsten.
+See the [remiw reward website](https://rewards.remix.ethereum.eth.limo) for a list of already minted badges.
```
Comment (Aniket-Engg) [inline README.md:16]: typo in `remiw`

File: README.md [hunk 6]
```diff
@@ -1,46 +1,35 @@
...
-npx hardhat verify --network ropsten DEPLOYED_CONTRACT_ADDRESS "Hello, Hardhat!"
-```
+When a trainer has been whitelisted in the contract (Please join our discord server and the community if you wish to be whitelisted as a trainer),
+he/she has the ability to allow student addresses to mint a soulbound nft, within the context of a workshop, classes, etc...
 
-# Performance optimizations
+The easiest way for a trainer to allow their students to mint a badge is to run the `grantRemixerMinting.ts` from the `scripts` folder.
+Then the student can either: 
+    - run the `remixerMint.ts` script.
+    - or browse the [remiw reward website](https://rewards.remix.ethereum.eth.limo) and go to the section `Mint a badge`
```
Comment (Aniket-Engg) [inline README.md:34]: same here

File: scripts/grantRemixerMinting.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+async function main() {
+    try {        
+        // proxy on Optimism.
+        const address = "0x5d470270e889b61c08C51784cDC73442c4554011"
+
+        // "signer" represents the current selected account and provider.
+        const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()
+
+        // getContractFactory retruns the compilation result.
```
Comment (Aniket-Engg) [inline scripts/grantRemixerMinting.ts:11]: typo in `retruns `

File: scripts/grantRemixerMinting.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        // This will be used in the next line to instantiate an "ethers.Contract" object ot interact with the contract.
+        const remixV1 = await ethers.getContractFactory("Remix")
+        
+        // we finally use the address, the contract interfact and the current context (provider and account)
```
Comment (Aniket-Engg) [inline scripts/grantRemixerMinting.ts:15]: typo in `interfact `

File: tests/test.ts (hunk 2 above)
Comment (Aniket-Engg) [reply tests/test.ts:237]: ah yes

**Review by Aniket-Engg — APPROVED** [review]
